File: sdk/go/chain.go
package tideline

import "fmt"

// ChainError is a record that does not verify, and where.
type ChainError struct {
	// Reason is one of: empty, not_run_started, seq_gap, prev_mismatch,
	// hash_mismatch. Shared with every other TLR/1 implementation.
	Reason string
	// Seq is the event at fault, or -1 when the record as a whole is at fault.
	Seq     int64
	Message string
}

func (e *ChainError) Error() string { return e.Message }

// ChainOK is what a valid record tells you.
type ChainOK struct {
	Len      int
	HeadSeq  uint64
	HeadHash string
	// Sealed is true when the record ends in run_finished.
	Sealed bool
}

// VerifyChain verifies an ordered run record end to end.
//
// It detects edited fields, inserted events, deleted events, and rewritten
// links, and names the event at fault. It cannot detect truncation of the tail —
// a valid prefix is a valid chain — which is what checkpoints are for. Check
// Sealed, and a checkpoint, before calling a record complete.
func VerifyChain(events []Event) (ChainOK, error) {
	if len(events) == 0 {
		return ChainOK{}, &ChainError{Reason: "empty", Seq: -1, Message: "the record is empty"}
	}
	if events[0].Kind != RunStarted {
		return ChainOK{}, &ChainError{
			Reason:  "not_run_started",
			Seq:     0,
			Message: fmt.Sprintf("a record must open with run_started, got %s", events[0].Kind),
		}
	}

	prev := ZeroHash
	for i := range events {
		e := &events[i]
		if e.Seq != uint64(i) {
			return ChainOK{}, &ChainError{
				Reason:  "seq_gap",
				Seq:     int64(i),
				Message: fmt.Sprintf("expected seq %d, found %d", i, e.Seq),
			}
		}
		if e.PrevHash != prev {
			return ChainOK{}, &ChainError{
				Reason:  "prev_mismatch",
				Seq:     int64(e.Seq),
				Message: fmt.Sprintf("event %d points at %s…, expected %s…", e.Seq, e.PrevHash[:16], prev[:16]),
			}
		}
		computed, err := e.ComputeHash(prev)
		if err != nil {
			return ChainOK{}, &ChainError{Reason: "prev_mismatch", Seq: int64(e.Seq), Message: err.Error()}
		}
		if computed != e.Hash {
			return ChainOK{}, &ChainError{
				Reason: "hash_mismatch",
				Seq:    int64(e.Seq),
				Message: fmt.Sprintf("event %d was altered: it hashes to %s…, but claims %s…",
					e.Seq, computed[:16], e.Hash[:16]),
			}
		}
		prev = computed
	}

	last := events[len(events)-1]
	return ChainOK{
		Len:      len(events),
		HeadSeq:  last.Seq,
		HeadHash: last.Hash,
		Sealed:   last.Kind == RunFinished,
	}, nil
}

File: sdk/ts/src/chain.ts
import type { TlrEvent } from "./event.js";
import { hashEvent } from "./canon.js";
import { ZERO_HEX } from "./sha256.js";

export type ChainFailure =
  | "empty"
  | "not_run_started"
  | "seq_gap"
  | "prev_mismatch"
  | "hash_mismatch";

/** A record that does not verify, and where. */
export class ChainError extends Error {
  constructor(
    readonly reason: ChainFailure,
    readonly seq: number | null,
    message: string,
  ) {
    super(message);
    this.name = "ChainError";
  }
}

export interface ChainOk {
  len: number;
  headSeq: number;
  headHash: string;
  /** True when the record ends in `run_finished`. */
  sealed: boolean;
}

/**
 * Verify an ordered run record end to end.
 *
 * Detects edited fields, inserted events, deleted events, and rewritten links,
 * and names the event at fault. It cannot detect truncation of the tail — a
 * valid prefix is a valid chain — which is what checkpoints are for. Check
 * `sealed`, and a checkpoint, before calling a record complete.
 */
export function verifyChain(events: TlrEvent[]): ChainOk {
  const first = events[0];
  if (!first) throw new ChainError("empty", null, "the record is empty");
  if (first.kind !== "run_started") {
    throw new ChainError(
      "not_run_started",
      0,
      `a record must open with run_started, got ${first.kind}`,
    );
  }

  let prev = ZERO_HEX;
  for (let i = 0; i < events.length; i++) {
    const e = events[i]!;
    if (e.seq !== i) {
      throw new ChainError("seq_gap", i, `expected seq ${i}, found ${e.seq}`);
    }
    if (e.prevHash !== prev) {
      throw new ChainError(
        "prev_mismatch",
        e.seq,
        `event ${e.seq} points at ${e.prevHash.slice(0, 16)}…, expected ${prev.slice(0, 16)}…`,
      );
    }
    const computed = hashEvent(e, prev);
    if (computed !== e.hash) {
      throw new ChainError(
        "hash_mismatch",
        e.seq,
        `event ${e.seq} was altered: it hashes to ${computed.slice(0, 16)}…, ` +
          `but claims ${e.hash.slice(0, 16)}…`,
      );
    }
    prev = computed;
  }

  const last = events[events.length - 1]!;
  return {
    len: events.length,
    headSeq: last.seq,
    headHash: last.hash,
    sealed: last.kind === "run_finished",
  };
}
